/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#pragma once

#include <memory>
#include <optional>
#include <string>

#include "chain/storage/proto/rocksdb_config.pb.h"
#include "chain/storage/storage.h"
#include "rocksdb/db.h"
#include "rocksdb/write_batch.h"

namespace resdb {
namespace storage {

std::unique_ptr<Storage> NewResRocksDB(
    const std::string& path, std::optional<RocksDBInfo> config = std::nullopt);
std::unique_ptr<Storage> NewResRocksDB(
    std::optional<RocksDBInfo> config = std::nullopt);

class ResRocksDB : public Storage {
 public:
  ResRocksDB(std::optional<RocksDBInfo> config_data = std::nullopt);
  virtual ~ResRocksDB();
  int SetValue(const std::string& key, const std::string& value) override;
  std::string GetValue(const std::string& key) override;
  int DelValue(const std::string& key) override;
  std::string GetAllValues(void) override;
  std::string GetRange(const std::string& min_key,
                       const std::string& max_key) override;

  int SetValueWithVersion(const std::string& key, const std::string& value,
                          int version) override;
  std::pair<std::string, int> GetValueWithVersion(const std::string& key,
                                                  int version) override;

  // Return a map of <key, <value, version>>
  std::map<std::string, std::pair<std::string, int>> GetAllItems() override;
  std::map<std::string, std::pair<std::string, int>> GetKeyRange(
      const std::string& min_key, const std::string& max_key) override;

  // Return a list of <value, version>
  std::vector<std::pair<std::string, int>> GetHistory(const std::string& key,
                                                      int min_version,
                                                      int max_version) override;
  std::vector<std::pair<std::string, int>> GetTopHistory(
      const std::string& key, int top_number) override;

  std::vector<std::string> GetKeysByPrefix(const std::string& prefix) override;

  std::vector<std::string> GetKeyRangeByPrefix(
      const std::string& start_prefix, const std::string& end_prefix) override;



  bool Flush() override;

 private:
  void CreateDB(const std::string& path);

 private:
  std::unique_ptr<::rocksdb::DB> db_ = nullptr;
  ::rocksdb::WriteBatch batch_;
  unsigned int num_threads_ = 1;
  unsigned int write_buffer_size_ = 64 << 20;
  unsigned int write_batch_size_ = 1;
};

}  // namespace storage
}  // namespace resdb